platform/linux-generic/pktio/socket_xdp.c

```diff
@@ -0,0 +1,676 @@
+	for (uint32_t i = 0U; i < pool->num + pool->skipped_blocks; ++i) {
+		hdr = packet_hdr(packet_from_event_hdr(event_hdr_from_index(pool, i)));
+
+		if (hdr->ms_pktio_idx == priv->sock_info.pktio_idx) {
+			hdr->ms_pktio_idx = 0U;
+			odp_packet_free(packet_handle(hdr));
+		}
+	}
+
+	return 0;
+}
+
+static inline void extract_data(const struct xdp_desc *rx_desc, uint8_t *pool_base_addr,
+				pkt_data_t *data)
+{
+	uint64_t frame_off;
+	uint64_t pkt_off;
+
+	/* UMEM "addresses" are offsets from start of a registered UMEM area.
+	 * Additionally, the packet data offset (where received packet data
+	 * starts within a UMEM frame) is encoded to the UMEM address with
+	 * XSK_UNALIGNED_BUF_OFFSET_SHIFT left bitshift when XDP_ZEROCOPY and
+	 * XDP_UMEM_UNALIGNED_CHUNK_FLAG are enabled. */
+	frame_off = rx_desc->addr;
+	pkt_off = xsk_umem__add_offset_to_addr(frame_off);
+	frame_off = xsk_umem__extract_addr(frame_off);
+	data->odp_hdr = xsk_umem__get_data(pool_base_addr, frame_off);
+	data->odp_pkt = packet_handle(data->odp_hdr);
+	data->pkt = xsk_umem__get_data(pool_base_addr, pkt_off);
+	data->len = rx_desc->len;
+}
+
+static uint32_t process_received(pktio_entry_t *pktio_entry, xdp_sock_info_t *sock_info,
+				 uint32_t start_idx, odp_packet_t packets[], int num)
+{
+	pkt_data_t pkt_data;
+	struct xsk_ring_cons *rx = &sock_info->rx;
+	uint8_t *base_addr = sock_info->umem_info->pool->base_addr;
+	const odp_proto_layer_t layer = pktio_entry->s.parse_layer;
+	const odp_proto_chksums_t in_chksums = pktio_entry->s.in_chksums;
+	const odp_pktin_config_opt_t opt = pktio_entry->s.config.pktin;
+	uint64_t l4_part_sum = 0U;
+	odp_pool_t *pool_hdl = &sock_info->umem_info->pool->pool_hdl;
+	odp_pktio_t pktio_hdl = pktio_entry->s.handle;
+	uint32_t num_rx = 0U;
+
+	for (int i = 0; i < num; ++i) {
+		extract_data(xsk_ring_cons__rx_desc(rx, start_idx++), base_addr, &pkt_data);
+		pkt_data.odp_hdr->ms_pktio_idx = 0U;
+		packet_init(pkt_data.odp_hdr, pkt_data.len);
+
+		if (layer) {
+			if (_odp_packet_parse_common(&pkt_data.odp_hdr->p, pkt_data.pkt,
+						     pkt_data.len, pkt_data.len,
+						     layer, in_chksums, &l4_part_sum, opt) < 0) {
+				odp_packet_free(pkt_data.odp_pkt);
+				continue;
+			}
+
+			if (pktio_cls_enabled(pktio_entry) &&
+			    _odp_cls_classify_packet(pktio_entry, pkt_data.pkt, pool_hdl,
+						     pkt_data.odp_hdr)) {
+				odp_packet_free(pkt_data.odp_pkt);
+				continue;
+			}
+		}
+
+		pkt_data.odp_hdr->seg_data = pkt_data.pkt;
+		pkt_data.odp_hdr->event_hdr.base_data = pkt_data.pkt;
+		pkt_data.odp_hdr->input = pktio_hdl;
+		packets[num_rx++] = pkt_data.odp_pkt;
+	}
+
+	return num_rx;
+}
+
+static int sock_xdp_recv(pktio_entry_t *pktio_entry, int index ODP_UNUSED, odp_packet_t packets[],
+			 int num)
+{
+	pkt_xdp_t *priv;
+	struct pollfd fd;
+	uint32_t start_idx = 0U, recvd, procd;
+
+	priv = pkt_priv(pktio_entry);
+	odp_ticketlock_lock(&priv->rx_lock);
+	fd.fd = xsk_socket__fd(priv->sock_info.xsk);
+	fd.events = POLLIN;
```

Comment:
This seems to be unnecessary if no wakeup is required.

Reply:
v11.